nova/virt/vmwareapi/vmops.py

```diff
@@ -2058,13 +2133,77 @@ def migrate_disk_and_power_off(self, context, instance, dest, flavor):
                                        total_steps=RESIZE_TOTAL_STEPS)
 
         # 1. Power off the instance
-        vm_util.power_off_instance(self._session, instance, vm_ref)
         self._update_instance_progress(context, instance,
                                        step=1,
                                        total_steps=RESIZE_TOTAL_STEPS)
+        vm_util.power_off_instance(self._session, instance, vm_ref)
+
+        # 2. Detach the volumes
+        self._update_instance_progress(context, instance,
+                                       step=2,
+                                       total_steps=RESIZE_TOTAL_STEPS)
+        self._detach_volumes(instance, block_device_info)
+
+        # 3. Detach network interfaces
+        self._update_instance_progress(context, instance,
+                                       step=3,
+                                       total_steps=RESIZE_TOTAL_STEPS)
+        device_change = self._get_network_device_change(
+            vm_ref, instance.image_meta, network_info, operation="remove")
+        vm_util.vm_device_change(self._session, vm_ref, device_change)
 
-    def confirm_migration(self, migration, instance, network_info):
+        # 4. Copy the VM to the dest
+        self._update_instance_progress(context, instance,
+                                       step=4,
+                                       total_steps=RESIZE_TOTAL_STEPS)
+        try:
+            self._do_migrate_disk(context, vm_ref, instance, dest_data)
+        except Exception as error:
+            # If the VM migration fails, we attempt to rollback what's been
+            # done so far and keep the VM in an ACTIVE state.
+            self._do_finish_revert_migration(
+                instance, block_device_info, network_info, vm_ref)
+            vm_util.power_on_instance(self._session, instance, vm_ref)
+            raise exception.InstanceFaultRollback(error)
```

Comment:
As I understand it, we will lose the traceback of our original exception here. Therefore, I would prefer to at least LOG it. Or does the `InstanceFaultRollback` exception do something more than convert the given object to a string for the exception message?

Edit: Additionally, do we need to add one of those with-statements saving exception context around our rollback?

Reply:
It seems to be logged and the original exception being reraised by the compute manager.

https://github.com/sapcc/nova/blob/daf10be0e68f451bd7987f46ebf7232201600ccd/nova/compute/manager.py#L8469-L8475